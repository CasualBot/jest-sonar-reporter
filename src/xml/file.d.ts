export declare const file: (testResult: any) => any;
