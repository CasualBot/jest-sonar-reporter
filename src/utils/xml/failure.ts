export const failure = (message: string): any => {
  // eslint-disable-next-line no-control-regex
  const filteredMessage = message.replace(/([\u001b]\[.{1,2}m)/g, '');
  const shortMessage = filteredMessage.replace(/[\n].*/g, '');
  return {
    failure: {
      _attr: {
        message: shortMessage
      },
      _cdata: filteredMessage
    }
  };
};
