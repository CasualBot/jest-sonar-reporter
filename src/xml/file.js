"use strict";
exports.__esModule = true;
exports.file = void 0;
var testCase_1 = require("./testCase");
var file = function (testResult) {
    //TODO: Make path relative or absolute based on config
    var aFile = [{ _attr: { path: testResult.testFilePath } }];
    var testCases = testResult.testResults.map(testCase_1.testCase);
    return { file: aFile.concat(testCases) };
};
exports.file = file;
