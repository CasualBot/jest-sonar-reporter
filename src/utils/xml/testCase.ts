import { failure } from "./failure";

export const testCase = (testResult: any): any => {
  let failures;
  const aTestCase = {
    _attr: {
      name: testResult.fullName || testResult.title,
      duration: testResult.duration || 0
    }
  }

  if (testResult.status === 'failed') {
    failures = testResult.failureMessages.map(failure)
    return {testCase: [aTestCase].concat(failures)}
  } else if (testResult.status === 'pending') {
    return {testCase: [aTestCase].concat({ skipped: {} } as any)};
  }
  return {testCase: aTestCase}
}
