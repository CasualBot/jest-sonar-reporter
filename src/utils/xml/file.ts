import { testCase } from './testCase';
import * as path from 'path';

export default (testResult: any, relativePath: boolean = false): any  => {
    let aFile: any;

    if(relativePath){
        aFile = [{_attr: { path: path.relative(process.cwd(), testResult.testFilePath) } }];
    } else {
        aFile = [{_attr: { path: testResult.testFilePath }}];
    }

    const testCases = testResult.testResults.map(testCase)

    return {file: aFile.concat(testCases)}
}