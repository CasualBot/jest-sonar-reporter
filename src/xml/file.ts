import { testCase } from './testCase';

export const file = (testResult: any): any  => {
    //TODO: Make path relative or absolute based on config
    const aFile = [{_attr: {path: testResult.testFilePath}}];
    const testCases = testResult.testResults.map(testCase)

    return {file: aFile.concat(testCases)}
}