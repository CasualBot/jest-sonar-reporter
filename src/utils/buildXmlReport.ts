import file from './xml/file';

export default (data: any, options: any = {}): any => {
  const aTestExecution = [{_attr: {version: '1'}}]
  const testResults = data.testResults.map((result: any) => { return file(result, options.relativePath) })

  return options?.formatForSonar56
    ? { unitTest: aTestExecution.concat(testResults) }
    : { testExecutions: aTestExecution.concat(testResults) };
}