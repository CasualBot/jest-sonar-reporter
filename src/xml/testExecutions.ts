import { file } from './file';

export const testExecutions = (data: any, formatForSonar56: boolean = false): any => {
  const aTestExecution = [{_attr: {version: '1'}}]
  const testResults = data.testResults.map(file)

  return formatForSonar56
    ? { unitTest: aTestExecution.concat(testResults) }
    : { testExecutions: aTestExecution.concat(testResults) };
}