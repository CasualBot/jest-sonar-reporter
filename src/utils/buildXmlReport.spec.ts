import xml from 'xml';
import buildXmlReport from './buildXmlReport';

describe('buildXmlReport', () => {
  test('root: <buildXmlReport version="1"> when not formatted for sonar 5.6.x', () => {
    const mock = { testResults: [] };

    const actualReport = xml(buildXmlReport(mock, false));

    expect(actualReport).toMatchSnapshot();
  });

  test('root: <unitTest version="1"> when formatted for sonar 5.6.x', () => {
    const mock = { testResults: [] };

    const actualReport = xml(buildXmlReport(mock, true));

    expect(actualReport).toMatchSnapshot();
  });

  test('file tag', () => {
    const mock = {
      testResults: [
        {
          testFilePath: 'test/FooTest.js',
          testResults: []
        },
        {
          testFilePath: 'test/BarTest.js',
          testResults: []
        }
      ]
    };

    const actualReport = xml(buildXmlReport(mock), true);

    expect(actualReport).toMatchSnapshot();
  });

  test('full report', () => {
    const mock = {
      testResults: [
        {
          testFilePath: 'test/FooTest.js',
          testResults: [
            {
              duration: 5,
              fullName: 'lorem ipsum'
            }
          ]
        },
        {
          testFilePath: 'test/BarTest.js',
          testResults: [
            {
              duration: 5,
              failureMessages: ['Lorem ipsum'],
              fullName: 'lorem ipsum',
              status: 'failed'
            }
          ]
        }
      ]
    };

    const actualReport = xml(buildXmlReport(mock), true);

    expect(actualReport).toMatchSnapshot();
  });
});
