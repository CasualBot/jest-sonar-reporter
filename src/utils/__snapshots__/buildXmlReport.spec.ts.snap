// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildXmlReport file tag 1`] = `
"<testExecutions version=\\"1\\">
    <file path=\\"test/FooTest.js\\">
    </file>
    <file path=\\"test/BarTest.js\\">
    </file>
</testExecutions>"
`;

exports[`buildXmlReport full report 1`] = `
"<testExecutions version=\\"1\\">
    <file path=\\"test/FooTest.js\\">
        <testCase name=\\"lorem ipsum\\" duration=\\"5\\"/>
    </file>
    <file path=\\"test/BarTest.js\\">
        <testCase name=\\"lorem ipsum\\" duration=\\"5\\">
            <failure message=\\"Lorem ipsum\\"><![CDATA[Lorem ipsum]]></failure>
        </testCase>
    </file>
</testExecutions>"
`;

exports[`buildXmlReport root: <buildXmlReport version="1"> when not formatted for sonar 5.6.x 1`] = `"<testExecutions version=\\"1\\"></testExecutions>"`;

exports[`buildXmlReport root: <unitTest version="1"> when formatted for sonar 5.6.x 1`] = `"<testExecutions version=\\"1\\"></testExecutions>"`;
